// @tolgee/react/dist/tolgee-react.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('react'), require('@tolgee/web')) :
    typeof define === 'function' && define.amd ? define(['exports', 'react', '@tolgee/web'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global["@tolgee/react"] = {}, global.React, global["@tolgee/web"]));
})(this, (function (exports, React, web) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

    const DEFAULT_REACT_OPTIONS = {
        useSuspense: true,
    };
    const TolgeeProviderContext = React__default["default"].createContext(undefined);
    const TolgeeProvider = ({ tolgee, options, children, fallback, }) => {
        const [loading, setLoading] = React.useState(!tolgee.isLoaded());
        React.useEffect(() => {
            tolgee.run().finally(() => {
                setLoading(false);
            });
            return () => {
                tolgee.stop();
            };
        }, [tolgee]);
        const optionsWithDefault = Object.assign(Object.assign({}, DEFAULT_REACT_OPTIONS), options);
        if (optionsWithDefault.useSuspense) {
            return (React__default["default"].createElement(TolgeeProviderContext.Provider, { value: { tolgee, options: optionsWithDefault } }, loading ? (fallback) : (React__default["default"].createElement(React.Suspense, { fallback: fallback || null }, children))));
        }
        return (React__default["default"].createElement(TolgeeProviderContext.Provider, { value: { tolgee, options: optionsWithDefault } }, loading ? fallback : children));
    };

    let globalContext;
    const GlobalContextPlugin = (options) => (tolgee) => {
        globalContext = {
            tolgee,
            options: Object.assign(Object.assign({}, DEFAULT_REACT_OPTIONS), options),
        };
        return tolgee;
    };
    function getGlobalContext() {
        return globalContext;
    }

    const useTolgeeContext = () => {
        const context = React.useContext(TolgeeProviderContext) || getGlobalContext();
        if (!context) {
            throw new Error("Couldn't find tolgee instance, did you forgot to use `TolgeeProvider`?");
        }
        return context;
    };

    const useRerender = () => {
        const [instance, setCounter] = React.useState(0);
        const rerender = React.useCallback(() => {
            setCounter((num) => num + 1);
        }, [setCounter]);
        return { instance, rerender };
    };

    const useTranslateInternal = (ns, options) => {
        const { tolgee, options: defaultOptions } = useTolgeeContext();
        const namespaces = web.getFallback(ns);
        const namespacesJoined = web.getFallbackArray(namespaces).join(':');
        const currentOptions = Object.assign(Object.assign({}, defaultOptions), options);
        // dummy state to enable re-rendering
        const { rerender, instance } = useRerender();
        const subscriptionRef = React.useRef();
        const subscriptionQueue = React.useRef([]);
        subscriptionQueue.current = [];
        const subscribeToNs = (ns) => {
            var _a;
            subscriptionQueue.current.push(ns);
            (_a = subscriptionRef.current) === null || _a === void 0 ? void 0 : _a.subscribeNs(ns);
        };
        const isLoaded = tolgee.isLoaded(namespaces);
        React.useEffect(() => {
            const subscription = tolgee.onNsUpdate(rerender);
            subscriptionRef.current = subscription;
            if (!isLoaded) {
                subscription.subscribeNs(namespaces);
            }
            subscriptionQueue.current.forEach((ns) => {
                subscription.subscribeNs(ns);
            });
            return () => {
                subscription.unsubscribe();
            };
        }, [isLoaded, namespacesJoined, tolgee]);
        React.useEffect(() => {
            tolgee.addActiveNs(namespaces);
            return () => tolgee.removeActiveNs(namespaces);
        }, [namespacesJoined, tolgee]);
        const t = React.useCallback((props) => {
            var _a;
            const fallbackNs = (_a = props.ns) !== null && _a !== void 0 ? _a : namespaces === null || namespaces === void 0 ? void 0 : namespaces[0];
            subscribeToNs(fallbackNs);
            return tolgee.t(Object.assign(Object.assign({}, props), { ns: fallbackNs }));
        }, [tolgee, instance]);
        if (currentOptions.useSuspense && !isLoaded) {
            throw tolgee.addActiveNs(namespaces, true);
        }
        return { t, isLoading: !isLoaded };
    };

    const useTranslate = (ns, options) => {
        const { t: tInternal, isLoading } = useTranslateInternal(ns, options);
        const t = React.useCallback((...params) => {
            // @ts-ignore
            const props = web.getTranslateProps(...params);
            return tInternal(props);
        }, [tInternal]);
        return { t, isLoading };
    };

    const wrapTagHandlers = (params) => {
        if (!params) {
            return undefined;
        }
        const result = {};
        Object.entries(params || {}).forEach(([key, value]) => {
            if (typeof value === 'function') {
                result[key] = (chunk) => {
                    return value(addReactKeys(chunk));
                };
            }
            else if (React__default["default"].isValidElement(value)) {
                const el = value;
                result[key] = (chunk) => {
                    return el.props.children !== undefined
                        ? React__default["default"].cloneElement(el)
                        : React__default["default"].cloneElement(el, {}, addReactKeys(chunk));
                };
            }
            else {
                result[key] = value;
            }
        });
        return result;
    };
    const addReactKeys = (val) => {
        if (Array.isArray(val)) {
            return React__default["default"].Children.toArray(val);
        }
        else {
            return val;
        }
    };

    const T = (props) => {
        const key = props.keyName || props.children;
        if (key === undefined) {
            // eslint-disable-next-line no-console
            console.error('T component: keyName not defined');
        }
        const defaultValue = props.defaultValue || (props.keyName ? props.children : undefined);
        const { t } = useTranslateInternal();
        const translation = addReactKeys(t({
            key: key,
            params: wrapTagHandlers(props.params),
            defaultValue,
            noWrap: props.noWrap,
            ns: props.ns,
        }));
        return React__default["default"].createElement(React__default["default"].Fragment, null, translation);
    };

    const useTolgee = (events) => {
        const { tolgee } = useTolgeeContext();
        const { rerender } = useRerender();
        React.useEffect(() => {
            const listeners = events === null || events === void 0 ? void 0 : events.map((e) => tolgee.on(e, rerender));
            return () => {
                listeners === null || listeners === void 0 ? void 0 : listeners.forEach((listener) => listener.unsubscribe());
            };
        }, [events === null || events === void 0 ? void 0 : events.join(':')]);
        return tolgee;
    };

    exports.GlobalContextPlugin = GlobalContextPlugin;
    exports.T = T;
    exports.TolgeeProvider = TolgeeProvider;
    exports.TolgeeProviderContext = TolgeeProviderContext;
    exports.useTolgee = useTolgee;
    exports.useTranslate = useTranslate;
    Object.keys(web).forEach(function (k) {
        if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
            enumerable: true,
            get: function () { return web[k]; }
        });
    });

    Object.defineProperty(exports, '__esModule', { value: true });

}));
